design/authenticators/authn_gcp/authn_gcp_solution_design.md

```diff
@@ -0,0 +1,313 @@
+    + [GCP authenticator flow](#gcp-authenticator-flow)
+  * [Backwards compatibility](#backwards-compatibility)
+  * [Performance](#performance)
+  * [Affected Components](#affected-components)
+- [Security](#security)
+- [Test Plan](#test-plan)
+- [Logs](#logs)
+  * [Audit](#audit)
+- [Documentation](#documentation)
+- [Open questions](#open-questions)
+- [Implementation plan](#implementation-plan)
+
+## Glossary
+
+| **Term** | **Description** |
+|----------|-----------------|
+|  GCP| Google Cloud Platform                 |
+|  Google compute engine (GCE)| Virtual machine in google (similar to AWS EC2)                  |
+|  Environment-provided service account| An identity for accessing private data on behalf of a service account inside Google Cloud environments |
+|  Service account key| App can self sign a google token with the service account key in order to accessing private data on behalf of a service account outside Google Cloud environments|
+
+
+
+## Useful links
+- [Feature doc](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+Authenticator#/)
+- [GCP authentication overview]( https://cloud.google.com/docs/authentication)
+- [Google metadata server](https://cloud.google.com/compute/docs/storing-retrieving-metadata#querying)
+
+## Issue description
+We want enable GCE hosts to authenticate Conjur with their google identity in order to retrieve secrets from Conjur.
+
+## Out of scope
+The only authentication strategy we are supporting is Environment-provided service account in google Compute engine.
+Other strategies are not supported: Service account key, OAuth 2.0 client, API key
+Other google components are not supported:  App Engine, GKE, Cloud Run, Cloud Functions
+
+## Solution
+We will add a new GCP authenticator to Conjur, which GCE hosts will can authenticate with by delivering their host Conjur identity and google identity as JWT token.
+
+As mentioned in the [feature doc](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Conjur+GCP+Authenticator#/), before any authentication 
+request is sent to Conjur, the admin will load the authenticator policy: 
+
+```yaml
+# policy id needs to match the convention `conjur/authn-azure/<service ID>`
+- !policy
+  id: conjur/authn-gcp/mygcp
+  body:
+  - !webservice
+      
+  - !group apps
+  
+  - !permit
+    role: !group apps
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+All the following configurations will have constants values and will not require any action from the customer:
+
+* **provider-uri** = `https://www.googleapis.com/oauth2/v1/certs` #represnts the address to fetch the public key in order to validate the JWT token
+* **mandatory-claims** = `iss, exp, aud, iat, nbf` #the list of the mandatory claims to validate in the JWT token
+* **iss** = `https://accounts.google.com` #the issuer claim we should validate in the JWT token
+* **aud** = `conjur` #the audience claim we should validate in the JWT token
```

Comment:
We may ask to inclide Conjur host in the `aud` claim.

Reply:
you should take it with Inbal it is not part of the requierments

Reply:
@eranha I am not sure i see an additional value here. can you elaborate how that that help?
If it is for security reasons I suggest to involve also Sasha

Reply:
@InbalZilberman @Tovli @sashaCher  @oburstein-hub 
By specifying the `http://conjur/account/host/my-policy/my-host/authenticate` in the audience claim same as you do in Conjure's authenticate URL `http://conjur/account/host/my-policy/my-host/authenticate` you are binding the token to its destination url, that is also what Google [recommends](https://cloud.google.com/compute/docs/instances/verifying-instance-identity).

`[AUDIENCE] is the unique URI agreed upon by both the instance and the system verifying the instance's identity. For example, the audience could be a URL for the connection between the two systems.`